### PR TITLE
Adapt `create_tag` client helper to reflect tag immutability

### PR DESCRIPTION
**Pull request by maxmynter**
Tags are immutable as per design from lakeFS (stated on lakeFS GUI tag site, no rename_tag API and their [docs](https://docs.lakefs.io/understand/model.html#tags)). The client helper returns the tag that is still pointing to the previously assigned commit. 

Therefore, 
- [x] set the `exist_ok` default of the `create_tag` client helper to `False` such that allowing the tag to exist (for e.g. idempotency of a script) is a conscious choice.
- [x] Log a warning that the tag was not reassigned if the tag already existed, to inform the user. 
- [x] Adapt the tutorial notebook to these changes
- [x] Add `delete_tag`  client helper such that explicit renaming by deletion and recreation is possible via our API
- [x] Create `delete_tag` tests 

Closes https://github.com/aai-institute/lakefs-spec/issues/165

- - -
This change is [Reviewable](https://reviewable.io/reviews/aai-institute/lakefs-spec/179)

### CODE REVIEW

File: src/lakefs_spec/client_helpers.py [hunk 1]
```diff
@@ -190,10 +190,27 @@ def create_tag(
         return client.tags_api.create_tag(repository=repository, tag_creation=tag_creation)
     except ApiException as e:
         if e.status == 409 and exist_ok:
+            logger.warning(f"The tag, '{tag}' already exists. It was not reassigned.")
```
Comment (nicholasjng) [inline src/lakefs_spec/client_helpers.py:193]: Make it short and simple, like `f"tag {tag} already exists and is immutable"`, and then match logs against `"tag .* already exists`.

**Review by nicholasjng — COMMENTED** [review]
I'm not sure what we settled on to fix the issue, but what do you say to the following behavior:

On HTTP409 (tag already exists): Get tag, check its `commit_id`, and verify it matches `rev_parse(client, repo, ref)`, otherwise, hard error.

This removes the ambiguity (and the need for `exist_ok`) entirely, and the fewest users will probably be saying "I want this tag and I don't care where it points", or am I mistaken?

So, in other words, allow existence ONLY if name and revision match.

File: docs/tutorials/demo_data_science_project.ipynb [hunk 2]
```diff
@@ -556,7 +556,7 @@
     "\n",
     "The above method for data versioning works great when you have experiment tracking tools to store and retrieve the commit SHA in automated pipelines. But it can be tedious to retrieve in manual prototyping. We can make selected versions of the dataset more accessible with semantic versioning by attaching a human-interpretable tag to a specific commit SHA.\n",
     "\n",
-    "Creating a tag is easiest when done inside a transaction, just like the files we already uploaded. To do this, simply call `tx.tag` on the transaction and supply the repository name, the commit SHA to tag, and the intended tag name."
+    "Creating a tag is easiest when done inside a transaction, just like the files we already uploaded. To do this, simply call `tx.tag` on the transaction and supply the repository name, the commit SHA to tag, and the intended tag name. The `exist_ok` flag handles creation errors, thrown if the tag does already exist. Since tags are immutable, the returned tag still points to the commit SHA it was previously assigned to. "
```
Comment (nicholasjng) [inline docs/tutorials/demo_data_science_project.ipynb:559]: Since `exist_ok` is now out of scope, I would propose this:

```suggestion
    "Creating a tag is easiest when done inside a transaction, just like the files we already uploaded. To do this, simply call `tx.tag` on the transaction and supply the repository name, the commit SHA to tag, and the intended tag name. Tags are immutable once created, so attempting to tag two different commits with the same name will result in an error."
```

File: src/lakefs_spec/client_helpers.py [hunk 3]
```diff
@@ -189,11 +185,29 @@ def create_tag(
     try:
         return client.tags_api.create_tag(repository=repository, tag_creation=tag_creation)
     except ApiException as e:
-        if e.status == 409 and exist_ok:
-            return client.tags_api.get_tag(repository=repository, tag=tag)
+        target_commit = rev_parse(client=client, repository=repository, ref=ref)
+        existing_tag = client.tags_api.get_tag(repository=repository, tag=tag)
+        if e.status == 409 and existing_tag.commit_id == target_commit.id:
+            return existing_tag
```
Comment (nicholasjng) [inline src/lakefs_spec/client_helpers.py:191]: For performance reasons, please move the rev-parse attempt in the `if e.status == 409` block so that it doesn't execute on e.g. a `PermissionError`:

```suggestion
        if e.status == 409:
                tagged_commit = rev_parse(client=client, repository=repository, ref=ref)
                existing_tag = client.tags_api.get_tag(repository=repository, tag=tag)
                if tagged_commit.id == existing_tag.commit_id:
                    return existing_tag
```

File: src/lakefs_spec/client_helpers.py [hunk 4]
```diff
@@ -189,11 +185,29 @@ def create_tag(
...
         raise e
 
 
+def delete_tag(client: LakeFSClient, repository: str, tag: str) -> None:
```
Comment (nicholasjng) [inline src/lakefs_spec/client_helpers.py:195]: Should we support `Ref` or `Commit` instances as tag inputs here, like e.g. in `rev_parse`?

File: tests/test_client_helpers.py [hunk 5]
```diff
@@ -4,37 +4,84 @@
...
-        existing_tag = client_helpers.create_tag(
-            client=fs.client, repository=repository, ref=temp_branch, tag=tag
+        commit_random_file_on_branch(
+            random_file_factory=random_file_factory,
+            fs=fs,
+            repository=repository,
+            temp_branch=temp_branch,
         )
-        assert new_tag == existing_tag
+        with pytest.raises(ApiException):
```
Comment (nicholasjng) [inline tests/test_client_helpers.py:60]: I assume that the error here will be a "tag already exists" error, have you verified this? What error message does this produce?

File: tests/util.py [hunk 6]
```diff
@@ -82,3 +85,17 @@ def make(self, fname: Optional[str] = None, size: int = 2**10) -> Path:
...
+    random_file_factory: RandomFileFactory, fs: LakeFSFileSystem, repository: str, temp_branch: str
+) -> None:
+    random_file = random_file_factory.make()
+    lpath = str(random_file)
+    rpath = f"{repository}/{temp_branch}/{random_file.name}"
+
+    fs.put(lpath, rpath, precheck=False)
+
+    commit = client_helpers.commit(
+        client=fs.client, repository=repository, branch=temp_branch, message="Commit File Factory"
```
Comment (nicholasjng) [inline tests/util.py:100]:
```suggestion
        client=fs.client, repository=repository, branch=temp_branch, message=f"Add file {random_file.name!r}"
```

See the other similar comment - can also be done in a follow-up if you prefer.

File: src/lakefs_spec/client_helpers.py (hunk 4 above)
Comment (maxmynter) [reply src/lakefs_spec/client_helpers.py:195]: Decided on just supporting `Ref` since the `Commit` object does not contain any reference to a tag. Plus, `list_tags` returns  a list of `Ref`s. Lastly, there could be edge cases such as multiple tags to a single commit (which should we then delete) and we want to avoid someone accidentally deleting someone elses tag.

File: tests/test_client_helpers.py (hunk 5 above)
Comment (maxmynter) [reply tests/test_client_helpers.py:60]: Btw, 
the terminal outputs the following:
```
raise ApiException(http_resp=r)
E           lakefs_sdk.exceptions.ApiException: (409)
E           Reason: Conflict
E           HTTP response headers: HTTPHeaderDict({'Content-Type': 'application/json', 'X-Content-Type-Options': 'nosniff', 'X-Request-Id': '19a26127-d216-4e93-af9b-45aff9e6a940', 'Date': 'Tue, 28 Nov 2023 19:08:34 GMT', 'Content-Length': '45'})
E           HTTP response body: {"message":"tag already exists: not unique"}

.venv/lib/python3.11/site-packages/lakefs_sdk/rest.py:237: ApiException
```

File: src/lakefs_spec/client_helpers.py [hunk 7]
```diff
@@ -189,11 +185,35 @@ def create_tag(
...
+        Name of the repository from which the tag will be deleted.
+    tag : str
+        Name of the tag to be deleted.
+    """
+    if isinstance(tag, str):
+        client.tags_api.delete_tag(repository=repository, tag=tag)
+    elif isinstance(tag, Ref):
+        client.tags_api.delete_tag(repository=repository, tag=tag.id)
+    else:
+        raise ValueError(r"Type of tag must be str or Ref.")
```
Comment (nicholasjng) [inline src/lakefs_spec/client_helpers.py:214]: This is unnecessary with a proper interface typing - after we tell users what is allowed, it's not our responsibility anymore if they pass something else.

I would go with this:

```suggestion
    if isinstance(tag, Ref):
        tag = tag.id

    client.tags_api.delete_tag(repository=repository, tag=tag.id)
```

File: tests/test_client_helpers.py (hunk 5 above)
Comment (nicholasjng) [reply tests/test_client_helpers.py:60]: Fair. Can you `assert e.status == 409`?

**Review by nicholasjng — APPROVED** [review]
Nice!
